plugins/badges-backend/src/lib/BadgeBuilder/DefaultBadgeBuilder.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2021 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { makeBadge, Format } from 'badge-maker';
+import { BadgeBuilder, BadgeOptions } from './types';
+import { Badge, BadgeFactories } from '../../types';
+
+export class DefaultBadgeBuilder implements BadgeBuilder {
+  constructor(private readonly factories: BadgeFactories) {}
+
+  public async getBadgeIds(): Promise<string[]> {
+    return Object.keys(this.factories);
+  }
+
+  public async createBadge(options: BadgeOptions): Promise<string> {
+    const factory = this.factories[options.badgeId];
+    const badge = factory
+      ? factory.createBadge(options.context)
+      : ({
+          label: 'unknown badge',
+          message: options.badgeId,
+          color: 'red',
+        } as Badge);
+
+    if (!badge) {
+      return '';
+    }
```

Comment:
Right, I'll refactor this to make the intent clearer. As it is now, the router returns a 404 in case the badge builder returns a falsey badge. But the badge factory could simply throw the `NotFoundError` directly itself instead, this also gets rid of the OptionalBadge, or Badge | undefined from the badge factory return type.

Reply:
Or is `InputError` more appropriate..?